src/main/java/org/folio/circulation/domain/MultipleRecords.java

```diff
@@ -63,8 +63,10 @@ public MultipleRecords(Collection<T> records, Integer totalRecords) {
   }
 
   public Map<String, T> toMap(Function<T, String> keyMapper) {
-    return getRecords().stream().collect(
-      Collectors.toMap(keyMapper, identity()));
+    return getRecords().stream() 
```

Comment:
Why was this change needed? How does the behaviour differ from the original?

Reply:
It was to address this exception that cropped up.

```
java.util.concurrent.ExecutionException:
java.lang.Exception: Expected Content Type: application/json Actual: text/plain; charset=ISO-8859-1 (Body: Duplicate key org.folio.circulation.domain.Location@4ea38988
java.lang.IllegalStateException: Duplicate key org.folio.circulation.domain.Location@4ea38988
        at java.util.stream.Collectors.lambda$throwingMerger$0(Collectors.java:133)
        at java.util.HashMap.merge(HashMap.java:1254)
        at java.util.stream.Collectors.lambda$toMap$58(Collectors.java:1320)
        at java.util.stream.ReduceOps$3ReducingSink.accept(ReduceOps.java:169)
        at java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1382)
        at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:482)
        at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:472)
        at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:708)
        at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
        at java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:499)
        at org.folio.circulation.domain.MultipleRecords.toMap(MultipleRecords.java:68)
        at org.folio.circulation.domain.LocationRepository.lambda$getLocations$6(LocationRepository.java:79)
        at org.folio.circulation.support.Result.lambda$24(Result.java:332)
        at org.folio.circulation.support.Result.next(Result.java:316)
        at org.folio.circulation.support.Result.map(Result.java:332)
        at org.folio.circulation.support.ResultBinding.lambda$0(ResultBinding.java:9)
        at java.util.concurrent.CompletableFuture.uniApply(CompletableFuture.java:602)
        at java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:577)
        at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:474)
        at java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:1962)
        at org.folio.circulation.support.CollectionResourceClient.lambda$18(CollectionResourceClient.java:214)
        at io.vertx.core.http.impl.HttpClientResponseImpl$BodyHandler.notifyHandler(HttpClientResponseImpl.java:296)
        at io.vertx.core.http.impl.HttpClientResponseImpl.lambda$bodyHandler$0(HttpClientResponseImpl.java:196)
        at io.vertx.core.http.impl.HttpClientResponseImpl.handleEnd(HttpClientResponseImpl.java:252)
        at io.vertx.core.http.impl.Http1xClientConnection$StreamImpl.endResponse(Http1xClientConnection.java:493)
        at io.vertx.core.http.impl.Http1xClientConnection$StreamImpl.access$000(Http1xClientConnection.java:194)
        at io.vertx.core.http.impl.Http1xClientConnection.handleResponseEnd(Http1xClientConnection.java:652)
        at io.vertx.core.http.impl.Http1xClientConnection.handleHttpMessage(Http1xClientConnection.java:602)
        at io.vertx.core.http.impl.Http1xClientConnection.handleMessage(Http1xClientConnection.java:560)
        at io.vertx.core.net.impl.ConnectionBase.handleRead(ConnectionBase.java:390)
        at io.vertx.core.impl.ContextImpl.executeTask(ContextImpl.java:320)
        at io.vertx.core.impl.WorkerContext.lambda$wrapTask$0(WorkerContext.java:34)
        at io.vertx.core.impl.TaskQueue.run(TaskQueue.java:76)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        at java.lang.Thread.run(Thread.java:748)
)
        at api.loans.LoanAPITests.queryLoans(LoanAPITests.java:1909)
        at api.loans.LoanAPITests.canGetPagedLoansWithItemsExceedingDefaultLimit(LoanAPITests.java:1858)
Caused by: java.lang.Exception:
Expected Content Type: application/json Actual: text/plain; charset=ISO-8859-1 (Body: Duplicate key org.folio.circulation.domain.Location@4ea38988
java.lang.IllegalStateException: Duplicate key org.folio.circulation.domain.Location@4ea38988
        at java.util.stream.Collectors.lambda$throwingMerger$0(Collectors.java:133)
        at java.util.HashMap.merge(HashMap.java:1254)
        at java.util.stream.Collectors.lambda$toMap$58(Collectors.java:1320)
        at java.util.stream.ReduceOps$3ReducingSink.accept(ReduceOps.java:169)
        at java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1382)
        at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:482)
        at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:472)
        at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:708)
        at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
        at java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:499)
        at org.folio.circulation.domain.MultipleRecords.toMap(MultipleRecords.java:68)
        at org.folio.circulation.domain.LocationRepository.lambda$getLocations$6(LocationRepository.java:79)
        at org.folio.circulation.support.Result.lambda$24(Result.java:332)
        at org.folio.circulation.support.Result.next(Result.java:316)
        at org.folio.circulation.support.Result.map(Result.java:332)
        at org.folio.circulation.support.ResultBinding.lambda$0(ResultBinding.java:9)
        at java.util.concurrent.CompletableFuture.uniApply(CompletableFuture.java:602)
        at java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:577)
        at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:474)
        at java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:1962)
        at org.folio.circulation.support.CollectionResourceClient.lambda$18(CollectionResourceClient.java:214)
        at io.vertx.core.http.impl.HttpClientResponseImpl$BodyHandler.notifyHandler(HttpClientResponseImpl.java:296)
        at io.vertx.core.http.impl.HttpClientResponseImpl.lambda$bodyHandler$0(HttpClientResponseImpl.java:196)
        at io.vertx.core.http.impl.HttpClientResponseImpl.handleEnd(HttpClientResponseImpl.java:252)
        at io.vertx.core.http.impl.Http1xClientConnection$StreamImpl.endResponse(Http1xClientConnection.java:493)
        at io.vertx.core.http.impl.Http1xClientConnection$StreamImpl.access$000(Http1xClientConnection.java:194)
        at io.vertx.core.http.impl.Http1xClientConnection.handleResponseEnd(Http1xClientConnection.java:652)
        at io.vertx.core.http.impl.Http1xClientConnection.handleHttpMessage(Http1xClientConnection.java:602)
        at io.vertx.core.http.impl.Http1xClientConnection.handleMessage(Http1xClientConnection.java:560)
        at io.vertx.core.net.impl.ConnectionBase.handleRead(ConnectionBase.java:390)
        at io.vertx.core.impl.ContextImpl.executeTask(ContextImpl.java:320)
        at io.vertx.core.impl.WorkerContext.lambda$wrapTask$0(WorkerContext.java:34)
        at io.vertx.core.impl.TaskQueue.run(TaskQueue.java:76)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        at java.lang.Thread.run(Thread.java:748)
)
```

Reply:
Do you know in what circumstances this error appeared?

Reply:
This is occurring when the number of ids/index names exceeds the batch size, causing it to break up the requests for items of the loans. Here is the records trying to be streamed. I have to dig in to find why there are duplicate locations being streamed when batching related record dereferencing.

``
{
  "primaryServicePointId" : "3b4648e7-edb1-4711-b31f-f8e7bd70edf7",
  "libraryName" : "Djanogly Learning Resource Centre",
  "campusName" : null,
  "institutionName" : null,
  "name" : "2nd Floor - Economics",
  "id" : "178924af-596a-4c1f-80bd-38c15d97e573"
}
{
  "primaryServicePointId" : "3b4648e7-edb1-4711-b31f-f8e7bd70edf7",
  "libraryName" : "Djanogly Learning Resource Centre",
  "campusName" : null,
  "institutionName" : null,
  "name" : "2nd Floor - Economics",
  "id" : "178924af-596a-4c1f-80bd-38c15d97e573"
}
```